docs: add security explanation page (ISD-5931) (#129)
canonical/content-cache-operator: docs/explanation/security.md

---
myst:
  html_meta:
    "description lang=en": "Explanation of the security design decisions in the Content Cache charm, including TLS, backend authentication, access control, and cached data risks."
---

(explanation_security)=

# Security

The charm makes deliberate security decisions at three boundaries: the connection from clients

> to nginx, the connection from nginx to backends, and the nginx process itself. This page

Reviewer: Nit, but the way you describe the boundaries here doesn't totally align with the section titles. I think it's fair to assume that "connection from clients to nginx" corresponds to the "Client to charm" section etc., but that mapping might not be immediately obvious to some users. 

I don't know whether we need to clarify anything, or if it's even accurate to rephrase to something like:
```suggestion
The charm makes deliberate security decisions at three boundaries: the connection from clients
to the charm through nginx, the connection from nginx to backends, and the nginx process itself. This page
```

I guess I'm wondering, do we need to explicitly write out the mapping between nginx and the charm itself? Or are they interchangeable for the purposes of this document?

Author: the charm handles the configuration of nginx, so the correct phrase will be client to charm and charm to backend, since the charm controls the nginx

certificate and private key. Without this relation, nginx serves all traffic over HTTP on
port 80 with no encryption.

The certificate and private key for each hostname are stored together in a single PEM file at
`/etc/nginx/certs/<hostname>.pem`. The file is owned by `www-data` with permissions `0o644`,
meaning it is readable by any local user on the machine. Operators should ensure the Juju
machine is not shared with untrusted local users.

### Client authentication

The charm provides no client authentication mechanism. Any client that can reach the charm's HTTP(S) listener (port 80 when no `certificates` relation is present, otherwise port 443) can request cached content. This is an intentional design constraint: the charm is built for publicly accessible static content. There are no plans to add a native authentication feature.

### Rate limiting

The charm does not configure nginx rate limiting
([`limit_req`](https://nginx.org/en/docs/http/ngx_http_limit_req_module.html) or
[`limit_conn`](https://nginx.org/en/docs/http/ngx_http_limit_conn_module.html)). Operators
who need protection against abuse or denial-of-service attacks must add rate limiting at a

> component placed in front of the charm, such as a load balancer, reverse proxy, or WAF.

Reviewer: Nit: Should we define WAF here? Will users already know about WAF, or is there a chance that they won't be familiar?

## Charm to backend

### Backend protocol

The `protocol` configuration option on `content-cache-backends-config` controls whether nginx

> contacts backends over HTTP or HTTPS. It defaults to `https`. Operators should keep this

Reviewer: ```suggestion
contacts backends over HTTP or HTTPS. The configuration defaults to `https`. Operators should keep this
```
Unless "it" means the charm. Either way, I think we should be more explicit here

default unless backends do not support HTTPS.

### Backend SSL certificate verification

The charm contacts backends over two separate code paths: the Lua healthcheck module (periodic
health pings) and the nginx `proxy_pass` directive (actual proxied requests). These have
different SSL verification behavior.

> **Healthchecks** — the `healthcheck-ssl-verify` option on `content-cache-backends-config`

Reviewer: ```suggestion
**Healthchecks** — the `healthcheck-ssl-verify` configuration option on `content-cache-backends-config`
```
nit

controls whether the Lua healthcheck module verifies the backend SSL certificate during health

> pings. It defaults to `true`. Setting it to `false` disables certificate verification for

Reviewer: ```suggestion
pings. The configuration defaults to `true`. Setting the configuration to `false` disables certificate verification for
```
Unless "it" means the charm (maybe it makes more sense that it's the charm that has defaults, but you set the configuration?)

> healthchecks and should only be used in controlled environments — for example, when backends

Reviewer: ```suggestion
healthchecks and should only be used in controlled environments, for example, when backends
```
nit

use self-signed certificates on a trusted private network.

**Proxied requests** — nginx's
[`proxy_ssl_verify`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_ssl_verify)
defaults to `off` and the charm does not override it. This means that even when
`protocol=https`, nginx encrypts the connection to the backend but does **not** verify the
SSL certificate of the backend. Traffic to the backend is protected against passive eavesdropping

> but not against a man-in-the-middle attack on that connection. There is no charm configuration

Reviewer: The inclusive language checker is complaining about "main-in-the-middle". Could we update to something like:
```suggestion
but not against a machine-in-the-middle attack on that connection. There is no charm configuration
```
Otherwise we'll need to place an ignore flag

option to enable SSL certificate verification for proxied backend connections.

## Internal

### Process user

On Ubuntu, the nginx package is configured to run worker processes as `www-data` via the
[`user` directive](https://nginx.org/en/docs/ngx_core_module.html#user) in the system
`nginx.conf`. `www-data` is a low-privilege system account with no login shell and no sudo
rights. If nginx were compromised, the attacker would have only the limited access of
`www-data` — they cannot read arbitrary files owned by other users or escalate privileges
without a separate exploit.

The charm creates and owns all cache and certificate files as `www-data`. The cache directory
at `/data/nginx/cache/` and the certificate files at `/etc/nginx/certs/` are owned by
`www-data` (set via `os.chown` in the charm code).

### Status page access

The nginx status page at `/nginx_status` and the backend health status page at

> `/nginx_backends_status` are restricted to `127.0.0.1` only using the nginx

Reviewer: ```suggestion
`/nginx_backends_status` are restricted to `127.0.0.1` using the nginx
```
Nit, I don't think "only" adds much value

[`allow`/`deny` directives](https://nginx.org/en/docs/http/ngx_http_access_module.html).

> External clients cannot access these endpoints. This is hardcoded in the generated nginx

Reviewer: ```suggestion
External clients cannot access these endpoints. This behavior is hard-coded in the generated nginx
```
I think we should be more explicit about what "this" is referring to. I think it's the behavior, but let me know if I misunderstood!

Also I think that we should hyphenate `hardcoded` instead, at least from the two style guides I read ([Salesforce](https://developer.salesforce.com/docs/atlas.en-us.salesforce_pubs_style_guide.meta/salesforce_pubs_style_guide/style_hard_coding.htm), [Microsoft](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/h/hard-code-hard-coded-hard-coding))

configuration and cannot be changed via charm configuration.

### Subprocess security

The charm uses `shell=False` for all subprocess calls, preventing shell injection attacks from
malformed configuration values.

## Cached data risks

### Public-only content

> The charm's cache key is based on `$scheme$proxy_host$request_uri` (the nginx default). It

Reviewer: ```suggestion
The charm's cache key is based on the nginx default `$scheme$proxy_host$request_uri`. It
```
nit, I think this flows a little more nicely

does not include request headers such as `Cookie` or `Authorization`. Two requests for the
same URL with different session cookies share a single cache entry: the first response is
cached and served to every subsequent requester of that URL, regardless of their identity.

Operators must ensure that only public, non-personalized content is routed through the charm.

> Routing personalized or session-dependent content through the charm will cause users to

Reviewer: ```suggestion
Routing personalized or session-dependent content through the charm means users will
```
nit, no strong opinion

receive each other's responses.

### No cache purge

The charm provides no built-in mechanism to purge a cached response on demand — for example,
if sensitive content was accidentally cached. Cache entries are removed when their
`proxy-cache-valid` TTL expires or when they are removed by nginx's default `inactive=10m`
behavior for `proxy_cache_path` (the charm does not override this). Operators who need on-demand purge capability must implement it directly at the nginx level

> outside the charm, or wait for natural expiry.

Reviewer: Is there a link to upstream docs on how to implement the on-demand purge capability?

Author: there are references of how this is done, for example https://github.com/FRiCKLE/ngx_cache_purge and https://www.f5.com/products/nginx/nginx-plus but these r third party/commercial solutions that's not available in the ubuntu nginx package currently

Reviewer 2: I would rephrase, these's no option to purge cache from backend perspective, only waiting for expiration

## Best practices

| Practice | Recommendation |
|---|---|

> | Enable TLS | Integrate with a `tls-certificates` provider charm |

Reviewer: ```suggestion
| TLS | Integrate with a `tls-certificates` provider charm to enable TLS |
```
Nit, to be consistent with the other rows (the first column is object-oriented rather than action-oriented)

| Backend protocol | Keep `protocol=https` (the default) |

> | Backend SSL verification | Keep `healthcheck-ssl-verify=true` (the default) |

Reviewer: ```suggestion
| Backend protocol | Keep the default `protocol=https` |
| Backend SSL verification | Keep the default `healthcheck-ssl-verify=true` |
```
nit

| Access control | Place an authenticating reverse proxy or WAF in front if the content is not fully public |
| Rate limiting | Add rate limiting at a component placed in front of the charm (load balancer, reverse proxy, or WAF) if abuse protection is needed |
| Cached content | Only route public, non-personalized content through the charm |

> | Machine access | Restrict local user access to the Juju machine — TLS private keys are readable by local users |

Reviewer: ```suggestion
| Machine access | Restrict local user access to the Juju machine |
```
Nit, to be consistent with the other rows (the other rows didn't provide more context beyond the recommendation). 

I think it's already mentioned that TLS private keys are readable by local users (in the `Transport security` section), so we should be okay to drop it here
